search/floydwarshall.go

```diff
@@ -5,226 +5,168 @@
 package search
 
 import (
-	"errors"
 	"math"
-	"sort"
+	"math/rand"
 
 	"github.com/gonum/graph"
+	"github.com/gonum/matrix/mat64"
 )
 
-// Finds all shortest paths between start and goal
-type AllPathFunc func(start, goal graph.Node) (path [][]graph.Node, cost float64, err error)
-
-// Finds one path between start and goal, which it finds is arbitrary
-type PathFunc func(start, goal graph.Node) (path []graph.Node, cost float64, err error)
-
-// This function returns two functions: one that will generate all shortest paths between two
-// nodes with ids i and j, and one that will generate just one path.
-//
-// This algorithm requires the CrunchGraph interface which means it only works on graphs with
-// dense node ids since it uses an adjacency matrix.
-//
-// This algorithm isn't blazingly fast, but is relatively fast for the domain. It runs at
-// O((number of vertices)^3) in best, worst, and average case, and successfully computes the cost
-// between all pairs of vertices.
-//
-// This function operates slightly differently from the others for convenience -- rather than
-// generating paths and returning them to you, it gives you the option of calling one of two
-// functions for each start/goal pair you need info for. One will return the path, cost, or an
-// error if no path exists.
-//
-// The other will return the cost and an error if no path exists, but it will also return ALL
-// possible shortest paths between start and goal. This is not too much more expensive than
-// generating one path, but it does obviously increase with the number of paths.
-func FloydWarshall(g graph.CrunchGraph, cost graph.CostFunc) (AllPathFunc, PathFunc) {
-	g.Crunch()
-	sf := setupFuncs(g, cost, nil)
-	successors, isSuccessor, cost, edgeTo := sf.successors, sf.isSuccessor, sf.cost, sf.edgeTo
-
-	nodes := denseNodeSorter(g.NodeList())
-	sort.Sort(nodes)
-	numNodes := len(nodes)
-
-	dist := make([]float64, numNodes*numNodes)
-	next := make([][]int, numNodes*numNodes)
-	for i := 0; i < numNodes; i++ {
-		for j := 0; j < numNodes; j++ {
-			if j != i {
-				dist[i+j*numNodes] = inf
-			}
-		}
-	}
-
-	for _, node := range nodes {
-		for _, succ := range successors(node) {
-			dist[node.ID()+succ.ID()*numNodes] = cost(edgeTo(node, succ))
-		}
+// FloydWarshall returns a shortest path tree for the graph g or false indicating
+// that a negative cycle exists in the graph. If weight is nil and the graph does not
+// implement graph.Coster, UniformCost is used.
+func FloydWarshall(g graph.Graph, weight graph.CostFunc) (paths ShortestPaths, ok bool) {
```

Comment:
Ideally, I'd like the CostFunc to go away too. It's the next thing on my shopping list at home. I think it should be dealt with by the graph interfaces.